tokio/src/runtime/mod.rs

```diff
@@ -172,6 +172,128 @@
 //! [`Builder::enable_io`]: crate::runtime::Builder::enable_io
 //! [`Builder::enable_time`]: crate::runtime::Builder::enable_time
 //! [`Builder::enable_all`]: crate::runtime::Builder::enable_all
+//!
+//! # Detailed runtime behavior
+//!
+//! This section gives more details into how the Tokio runtime will schedule
+//! tasks for execution.
+//!
+//! At its most basic level, a runtime has a collection of tasks that need to be
+//! polled. It will repeatedly remove a task from that collection and poll it.
+//! When the collection is empty, the thread will go to sleep until a task is
+//! added to the collection.
+//!
+//! However, the above is not sufficient to guarantee a well-behaved runtime.
+//! For example, the runtime might have a single task that is always ready to be
+//! polled, and poll that task every time. This is a problem because it starves
+//! other tasks by not polling them. To solve this, Tokio provides the following
+//! fairness guarantee:
+//!
+//! > Under the following two assumptions:
+//! >
+//! > * There is some number `MAX_TASKS` such that the total number of tasks on
+//! >   the runtime never exceeds `MAX_TASKS`.
+//! > * There is some number `MAX_POLL` such that calling [`poll`] on any task
+//! >   spawned on the runtime returns within `MAX_POLL` time units.
+//! >
+//! > Then, it is guaranteed that tasks are scheduled fairly, that is, there is
+//! > some number `MAX_DELAY` such that when a task becomes ready, it will be
+//! > polled by the runtime within `MAX_DELAY` time units.
+//!
+//! Other than the above fairness guarantee, there is no guarantee about the
+//! order in which tasks are polled. There is also no guarantee that the runtime
+//! is equally fair to all tasks. For example, if the runtime has two tasks A
+//! and B that are both ready, then the runtime may poll A five times before it
+//! polls B. This is the case even if A yields using [`yield_now`]. All that is
+//! guaranteed is that it will poll B eventually.
+//!
+//! Normally, tasks are polled only if they have notified the runtime by waking
+//! their waker. However, this is not guaranteed, and Tokio may poll tasks that
+//! have not notified the runtime under some circumstances. This is called a
+//! spurious wakeup.
+//!
+//! ## IO and timers
+//!
+//! Beyond just polling tasks, the runtime must also manage IO resources and
+//! timers. It does this by periodically checking whether there are any IO
+//! resources or timers that are ready, and notifying the relevant task so that
+//! it will be polled.
+//!
+//! These checks are performed periodically between polling tasks. Under the
+//! same assumptions as the previous fairness guarantee, Tokio guarantees that
+//! it will notify tasks with an IO or timer event within some maximum number of
+//! time units.
+//!
+//! ## Current behavior (current thread runtime)
```

Comment:
Reading this title felt to me like it implied that there would be a "future behavior" section as well.

Given that the first paragraph calls out that this behavior may change, this title could be simplied to "Current thread runtime behavior". The same for the multi thread runtime heading below.

Reply:
By using the word "current", I meant to emphasize that this section only describes how it works today, and not necessarily in the future.

Reply:
How about "Behavior at the time of writing"?